File: android/app/src/main/java/de/psmobile/ui/OpenSourceView.kt
package de.psmobile.ui

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import de.psmobile.shared.rules.OpenSource
import de.psmobile.ui.theme.PrusaColors

/// "Open Source & Lizenzen": woraus Slicer Mobile besteht und wo der
/// Quellcode liegt. Gegenstueck zu `OpenSourceView.swift`.
///
/// PrusaSlicer steht unter AGPL-3.0 - ein abgeleitetes Werk muss seinen
/// Quellcode zugaenglich machen und die verwendeten Bibliotheken nennen.
/// Die Liste selbst kommt aus `OpenSource` im gemeinsamen Modul.
@Composable
fun OpenSourceView(onClose: () -> Unit) {
    val ps = LocalPsScale.current
    val browser = LocalUriHandler.current
    val deutsch = Sprache.aktuell == "de"

    BackHandler(onBack = onClose)

    Box(
        Modifier
            .fillMaxSize()
            .background(PrusaColors.background),
    ) {
        Column(Modifier.fillMaxSize(), horizontalAlignment = Alignment.Start) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(ps.touch(56))
                    .padding(horizontal = ps.pt(16)),
                horizontalArrangement = Arrangement.spacedBy(ps.pt(16)),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    "‹  " + st("Back", "Zurück"),
                    Modifier
                        .clickable(onClick = onClose)
                        .testTag("opensource.zurueck"),
                    fontSize = ps.font(15), color = PrusaColors.orange,
                )
                Text(st("Open source & licenses", "Open Source & Lizenzen"), fontSize = ps.font(20), color = PrusaColors.textPrimary)
                Spacer(Modifier.weight(1f))
            }
            HorizontalDivider(color = PrusaColors.divider)

            Column(
                Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
                    .padding(ps.pt(16))
                    .testTag("opensource.liste"),
                verticalArrangement = Arrangement.spacedBy(ps.pt(10)),
            ) {
                Text(
                    st(
                        "Slicer Mobile is an unofficial app built on PrusaSlicer. It is not affiliated with or endorsed by Prusa Research. " +
                            "Like PrusaSlicer it is released under the ${OpenSource.LIZENZ} license – the complete source code is public.",
                        "Slicer Mobile ist eine inoffizielle App auf Basis von PrusaSlicer. Sie ist nicht mit Prusa Research verbunden und nicht von Prusa Research freigegeben. " +
                            "Wie PrusaSlicer steht sie unter der Lizenz ${OpenSource.LIZENZ} – der vollständige Quellcode ist öffentlich.",
                    ),
                    fontSize = ps.font(13), color = PrusaColors.textPrimary,
                )

                Box(
                    Modifier
                        .fillMaxWidth()
                        .heightIn(min = ps.touch(48))
                        .clip(RoundedCornerShape(ps.pt(3)))
                        .background(PrusaColors.orange)
                        .clickable { runCatching { browser.openUri(OpenSource.REPO) } }
                        .testTag("opensource.github")
                        .padding(horizontal = ps.pt(14)),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        st("Source code on GitHub", "Quellcode auf GitHub"),
                        fontSize = ps.font(14), fontWeight = FontWeight.SemiBold, color = Color.White,
                    )
                }
                Text(OpenSource.REPO, fontSize = ps.font(11), color = PrusaColors.textMuted)

                Text(
                    st("Third-party components", "Verwendete Komponenten").uppercase(),
                    Modifier.padding(top = ps.pt(12)),
                    fontSize = ps.font(11), fontWeight = FontWeight.SemiBold, color = PrusaColors.textMuted,
                )
                OpenSource.fuer(OpenSource.Plattform.ANDROID).forEachIndexed { index, k ->
                    Column(
                        Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(ps.pt(10)))
                            .background(PrusaColors.panelRaised)
                            .clickable { runCatching { browser.openUri(k.url) } }
                            .testTag("opensource.komponente.$index")
                            .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
                        verticalArrangement = Arrangement.spacedBy(ps.pt(2)),
                    ) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(k.name, Modifier.weight(1f), fontSize = ps.font(14), color = PrusaColors.textPrimary)
                            Text(k.lizenz, fontSize = ps.font(11), color = PrusaColors.orange)
                        }
                        Text(if (deutsch) k.rolleDe else k.rolleEn, fontSize = ps.font(11), color = PrusaColors.textMuted)
                        Text(k.url, fontSize = ps.font(10), color = PrusaColors.textMuted)
                    }
                }
                Spacer(Modifier.size(ps.pt(8)))
            }
        }
        PSMarke(name = "opensource")
    }
}

File: ios/PSMobile/Screens/OpenSourceView.swift
import SwiftUI
import PSMShared

/// "Open Source & Lizenzen": woraus Slicer Mobile besteht und wo der
/// Quellcode liegt. Gegenstueck zu `OpenSourceView.kt`.
///
/// PrusaSlicer steht unter AGPL-3.0 - ein abgeleitetes Werk muss seinen
/// Quellcode zugaenglich machen und die verwendeten Bibliotheken nennen.
/// Die Liste selbst kommt aus `OpenSource` im gemeinsamen Modul.
struct OpenSourceView: View {

    var onClose: () -> Void

    @Environment(\.psScale) private var ps
    @Environment(\.openURL) private var openURL

    private var deutsch: Bool { Lang.shared.current == "de" }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            kopfzeile
            Divider().background(PrusaColors.divider)
            ScrollView {
                VStack(alignment: .leading, spacing: ps.pt(10)) {
                    Text(st(
                        "Slicer Mobile is an unofficial app built on PrusaSlicer. It is not affiliated with or endorsed by Prusa Research. " +
                            "Like PrusaSlicer it is released under the \(OpenSource.shared.LIZENZ) license – the complete source code is public.",
                        "Slicer Mobile ist eine inoffizielle App auf Basis von PrusaSlicer. Sie ist nicht mit Prusa Research verbunden und nicht von Prusa Research freigegeben. " +
                            "Wie PrusaSlicer steht sie unter der Lizenz \(OpenSource.shared.LIZENZ) – der vollständige Quellcode ist öffentlich."))
                        .font(.system(size: ps.font(13)))
                        .foregroundStyle(PrusaColors.textPrimary)
                        .fixedSize(horizontal: false, vertical: true)

                    Button {
                        if let url = URL(string: OpenSource.shared.REPO) { openURL(url) }
                    } label: {
                        Text(st("Source code on GitHub", "Quellcode auf GitHub"))
                            .font(.system(size: ps.font(14), weight: .semibold))
                            .foregroundStyle(.white)
                            .frame(maxWidth: .infinity)
                            .frame(minHeight: ps.touch(48))
                            .background(PrusaColors.orange)
                            .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                            .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityIdentifier("opensource.github")
                    Text(OpenSource.shared.REPO)
                        .font(.system(size: ps.font(11)))
                        .foregroundStyle(PrusaColors.textMuted)

                    Text(st("Third-party components", "Verwendete Komponenten").uppercased())
                        .font(.system(size: ps.font(11), weight: .semibold))
                        .foregroundStyle(PrusaColors.textMuted)
                        .padding(.top, ps.pt(12))
                    let komponenten = OpenSource.shared.fuer(plattform: OpenSource.Plattform.ios)
                    ForEach(Array(komponenten.enumerated()), id: \.offset) { index, k in
                        Button {
                            if let url = URL(string: k.url) { openURL(url) }
                        } label: {
                            VStack(alignment: .leading, spacing: ps.pt(2)) {
                                HStack {
                                    Text(k.name)
                                        .font(.system(size: ps.font(14)))
                                        .foregroundStyle(PrusaColors.textPrimary)
                                    Spacer()
                                    Text(k.lizenz)
                                        .font(.system(size: ps.font(11)))
                                        .foregroundStyle(PrusaColors.orange)
                                }
                                Text(deutsch ? k.rolleDe : k.rolleEn)
                                    .font(.system(size: ps.font(11)))
                                    .foregroundStyle(PrusaColors.textMuted)
                                Text(k.url)
                                    .font(.system(size: ps.font(10)))
                                    .foregroundStyle(PrusaColors.textMuted)
                            }
                            .padding(.horizontal, ps.pt(14))
                            .padding(.vertical, ps.pt(10))
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(PrusaColors.panelRaised)
                            .clipShape(RoundedRectangle(cornerRadius: ps.pt(10)))
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .accessibilityIdentifier("opensource.komponente.\(index)")
                    }
                    Spacer().frame(height: ps.pt(8))
                }
                .padding(ps.pt(16))
            }
            .accessibilityIdentifier("opensource.liste")
        }
        .background(PrusaColors.background)
        .overlay(alignment: .topLeading) { PSMarke(name: "opensource") }
    }

    private var kopfzeile: some View {
        HStack(spacing: ps.pt(16)) {
            Button(action: onClose) {
                Text("‹  " + st("Back", "Zurück"))
                    .font(.system(size: ps.font(15)))
                    .foregroundStyle(PrusaColors.orange)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("opensource.zurueck")

            Text(st("Open source & licenses", "Open Source & Lizenzen"))
                .font(.system(size: ps.font(20)))
                .foregroundStyle(PrusaColors.textPrimary)
            Spacer()
        }
        .padding(.horizontal, ps.pt(16))
        .frame(height: ps.touch(56))
    }

    private func st(_ english: String, _ german: String) -> String {
        SimpleModeState.shared.text(english: english, german: german)
    }
}
